### ISSUE TITLE
Questions about sstable

### ISSUE THREAD

**Issue by lovebaihezi** [bokket/Kirsidb#1]
[Related Code](https://github.com/bokket/Kirsidb/blob/dd19dddd9c6ea0292fb6e63d726793ab182750ba/sstable/block_iter.cpp#L125)

# Questions

**Comment by bokket** [comment]
In order to reduce data storage, each Key-Value is not stored independently in SST, but the concept of shared key is introduced.
Each Key-Value is stored in the format of :

<!DOCTYPE html><meta http-equiv="content-type" content="text/html; charset=utf-8">
shared_length | var_4 byte
-- | --
unshared_length | var_4 byte
value_length | var_4 byte
delta_key | string
value | string

```shared_length    := 0
unshared_length  := len(key)
value_length     := len(value)
delta_key        := key
value            := value
```

The contents of `cur_key` must be cleared when the iterator is incremented. Or will get a duplicate `cur_key` next time.
But once `cur_key` is cleared ,`last_key` at [init() ](https://github.com/bokket/Kirsidb/blob/dd19dddd9c6ea0292fb6e63d726793ab182750ba/sstable/block_iter.cpp#L172)will share the original pointer with `cur_key` and will be cleared together.
Move `std::string::data` to `last_key`:hand over ownership of `data` to `last_key`.
The life cycle of the `cur_key` must finish `init()` or `Key()` and destory in `increment()`.

**Comment by lovebaihezi** [comment]
Can we move the cur_key_ to a local string?

```cpp
std::string cur = std::move(cur_key_);
```

which the compiler gonna help you empty the content of the `cur_key_`

And in the end, it is obviously that `cur_key_`, `tmp` will destroy after the `init` function call, and the content of the `cur_key_` will finally store into `last_key`, why can't we just move the content from `cur_key_` to `last_key`, and construct `string_view` from `last_key_`?

**Comment by bokket** [comment]
```
void BlockConstIter::increment() {
    std::string_view tmp=std::move(cur_key_);
    last_key_=tmp.data();
}
```
Look at the test that it share the same address.

```
void BlockConstIter::increment() {
    last_key_=std::move(cur_key_);
    last_key_=std::string_view{last_key_};
}
```
It feels the same way as above because `last_key_` and `cur_key_` type is string.Need constructor.

**Comment by bokket** [comment]
can work.coredump in case
```
27)[key]:}6dgP>M>%{aS$s&y		[value]:`);n&n6ec.2_HQy}-
 increment last_key_: }6dgP>M>%{aS$s&y
 increment: }6dgP>M>%{aS$s&y`);n&n6ec.2_HQy}-         HVTz,HV~U<=c+_fS7cn_11 | 0 | cur_len:34
 len:34 | shared_key_len_:1 | unshared_key_len_:3 | value_len_:19 |
 last_key: \06dgP>M>%{aS$s&y| cur_entry:HVTz,HV~U<=c+_fS7cn_11 | cur_key:

Expected equality of these values:
  it.key()
    Which is: "\0HVT"
  it2->first
    Which is: "}HVT"
```

**Comment by lovebaihezi** [comment]
I mean

```cpp
this->last_key_ = std::move(this->cur_key_);
```

Or

```cpp
std::swap(this->last_key_, this->cur_key_)
```

**Comment by bokket** [comment]
Useful.Change related code[#179](https://github.com/bokket/Kirsidb/blob/dd19dddd9c6ea0292fb6e63d726793ab182750ba/sstable/block_iter.cpp#L179).
```
void BlockConstIter::init(const char *p) {
  //cur_key_
        if(last_key_!= nullptr) {
            LOG_INFO("last_key:{}| cur_entry:{} | cur_key:{}",last_key_,cur_entry_,cur_key_);
        }

        assert(cur_key_.empty());

        cur_key_.reserve(shared_key_len_+unshared_key_len_);
        cur_key_.append(last_key_.data(),shared_key_len_);
        cur_key_.append(cur_entry_.data(),unshared_key_len_);
}
```
Need profiler compare.

**Comment by bokket** [comment]
fix in [#50ef672](https://github.com/bokket/Kirsidb/commit/50ef672d1107c173d6b04168f4f4c04498ce5a31)
